## Fuzzy Matching and Fuzzy Pandas
      
[Max Harlow](https://twitter.com/maxharlow), a journalist at the Financial Times, wrote this library `csvmatch`, and he's been adding new algorithms to facilitate fuzzy matching across datasets. He's used it for a bunch of stories, including:
- https://www.theguardian.com/uk-news/2014/jul/09/offshore-tax-dealings-celebrities-sportsmen-leaked-jersey-files
- https://www.theguardian.com/politics/2014/jul/08/offshore-secrets-wealthy-political-donors

Similar techniques have also been used in other stories like:
- https://www.globalwitness.org/en/campaigns/oil-gas-and-mining/myanmarjade
- https://www.irinnews.org/investigation/2016/09/02/exclusive-un-paying-blacklisted-diamond-company-central-african-republic

But, wait, first: 

### What is Fuzzy Matching? 

Automating the look-up for names in documents is [inherently imprecise](https://www.elastic.co/blog/found-fuzzy-search). The computer can't _know_ that different representations of the same _thing_ refer to the same _thing_. For example: 
- _Apple Inc._; _Apple Computer Company_; _Apple Computer, Inc._; and _Apple_ all refer to the fruit company. 
- _Samuel Langhorne Clemens_, _Samuel L. Clemens_, _Samuel Clemens_ and Mark Twain all refer to the same person. 
- _Robert Ford_, _Rob Ford_, and _Robert Frod_ refer to the same person **probably**. 

When you're working with unstructured data, you can't take anything for granted. Least of all, you can't assume that:
- documents will have correct spellings
- first, last, and middle names will exist in all documents
- the abbreviated/shortened names of people won't make an appearance (e.g. Jon instead of Jonathan, Tom instead of Thomas, Phil instead of Philip, etc.) 

So, when you're living in an uncertain world, you try to make things slightly more _certain_ with **Fuzzy Matching**. You might not hit 100 percent, but at least you'll hit more than what you would without fuzzy matching. 

There are multiple algorithms that try to minimise the uncertainty/enable fuzzy matching. The library we are going to be look at today incorporates a bunch of these, instead of just doing one thing. 

This notebook's predominantly based on an [awesome NICAR2019 presentation](https://docs.google.com/presentation/d/1djKgqFbkYDM8fdczFhnEJLwapzmt4RLuEjXkJZpKves/) where Max Harlow (the aforementioned news app developer at the Financial Times) demonstrated [csvmatch](https://github.com/maxharlow/csvmatch). And, then, Soma basically created a library to make it with Pandas. 

Worth remembering that there are no shortcuts in life, and few panaceas. Depending on the project you're working on, you might be more inclined to use one algorithm or the other. Or, you know, try a few of them and see what happens. And, also, remember: all computational tools you use need to hand-in-hand with traditional reporting. People share names, there's more than one John Smith, etc. 


In [1]:
# Make sure you `pip install fuzzy_pandas` first. 

import pandas as pd
import fuzzy_pandas as fpd

### A Toy Example

We'll be working with two toy datasets first, just to get going and get an idea as to what's possible. The names of the files are not terribly imaginative: `data1.csv` and `data2.csv`. And, they both contain structured data: names, code names and locations of characters from John le Carré's spy thriller: Tinker Tailor Soldier Spy. 

Right, let's have a look. 


In [2]:
df1 = pd.read_csv("sources/data1.csv")
df2 = pd.read_csv("sources/data2.csv")

In [3]:
df1

name  location   codename
0   George Smiley    London  Beggerman
1  Percy Alleline    London     Tinker
2       Roy Bland    London    Soldier
3  Toby Esterhase    Vienna    Poorman
4   Peter Guillam   Brixton       none
5     Bill Haydon    London     Tailor
6    Oliver Lacon    London       none
7    Jim Prideaux  Slovakia       none
8    Connie Sachs    Oxford       none

In [4]:
df2

Person Name Location
0  Maria Andreyevna Ostrakova   Russia
1                Otto Leipzig  Estonia
2               George SMILEY   London
3               Peter Guillam  Brixton
4                  Konny Saks   Oxford
5                Saul Enderby   London
6                 Sam Collins  Vietnam
7              Tony Esterhase   Vienna
8           Claus Kretzschmar  Hamburg

### Exact matches

We start with doing "exact matches", i.e. both tables should have the exact same name. Capitalisation matters, accents matter. With this function, for example:
- John le Carre will not match with John le Carré
- George SMILEY will not match with George Smiley

Based on what you see in the data frames above, how many matches do you expect? 

In [5]:
fpd.fuzzy_merge(df1, df2, left_on='name', right_on='Person Name')

name location codename    Person Name Location
0  Peter Guillam  Brixton     none  Peter Guillam  Brixton

Right, so, we only find one match as expected. But, are there any other matches that a _smarter_ algorithm could find? Let's try something called **Levenshtein**, a nifty simple algorithm that's pretty common. It's the basis for a bunch of spellcheck algorithms, amongst other things, and the way it works is it checks the number of characters that are different between two inputs, and if the _distance_ is small enough, it assumes the two words are the same. 

For example, in the above two data frames, you have Toby Esterhase and Tony Esterhase, which means the Levenshtein distance is 1 (The 'b' v. 'n' in To(b,n)y.). 

In [6]:
fpd.fuzzy_merge(df1, df2, left_on='name', right_on='Person Name', method='levenshtein')

name location   codename     Person Name Location
0   George Smiley   London  Beggerman   George SMILEY   London
1  Toby Esterhase   Vienna    Poorman  Tony Esterhase   Vienna
2   Peter Guillam  Brixton       none   Peter Guillam  Brixton

The other thing you'll notice above is that, by default, the _Levenshtein_ algorithm doesn't care about case. 

However, are we still missing potential matches? 

When we work with any algorithms, we need a confidence threshold that we decide on. By default, the `csvmatch` algorithm has a `threshold` of 0.6, i.e. only if the algorithm returns a match score greater than or equal to 0.6 will it return a match. 

The score, in this case, is calculated using the below formula: 

> `1 - (distance/maximum(value1, value2))`

We can be slightly more conservative with the threshold, and we get a Brand New Result in our output. 

In [7]:
fpd.fuzzy_merge(df1, df2, left_on='name', right_on='Person Name', method='levenshtein', threshold=0.55)

name location   codename     Person Name Location
0   George Smiley   London  Beggerman   George SMILEY   London
1  Toby Esterhase   Vienna    Poorman  Tony Esterhase   Vienna
2   Peter Guillam  Brixton       none   Peter Guillam  Brixton
3    Connie Sachs   Oxford       none      Konny Saks   Oxford

This is _cool_. By changing the threshold, we found another match based on what the pronunciation of the two names are: Connie and Konny. **But, what could be cooler?**

In [8]:
fpd.fuzzy_merge(df1, df2, left_on='name', right_on='Person Name', method='metaphone')

name location   codename    Person Name Location
0  George Smiley   London  Beggerman  George SMILEY   London
1  Peter Guillam  Brixton       none  Peter Guillam  Brixton
2   Connie Sachs   Oxford       none     Konny Saks   Oxford

The **metaphone** algorithm does phonetic matching, and gives you results based on that. 

Note: In theory, the documentation says that you can combine a couple of these algorithms if you're so inclined. But, it looks like when you combine two algorithms, it doesn't _quite_ work. ¯\_(ツ)_/¯

In [9]:
fpd.fuzzy_merge(df1, df2, left_on='name', right_on='Person Name', method=['levenshtein', 'metaphone'])

name location   codename     Person Name Location
0   George Smiley   London  Beggerman   George SMILEY   London
1  Toby Esterhase   Vienna    Poorman  Tony Esterhase   Vienna
2   Peter Guillam  Brixton       none   Peter Guillam  Brixton

In [10]:
## swap the methods around and then look at the results, too.
fpd.fuzzy_merge(df1, df2, left_on='name', right_on='Person Name', method=['metaphone','levenshtein'])

name location   codename    Person Name Location
0  George Smiley   London  Beggerman  George SMILEY   London
1  Peter Guillam  Brixton       none  Peter Guillam  Brixton
2   Connie Sachs   Oxford       none     Konny Saks   Oxford

What do you think is happening here? 

This is important—you're often going to be using tools built by other folks, but where there's code, there are bugs. You should make sure that you play with the tool a bit to make sure it's doing _exactly_ what you think it's doing. And, if it's not, you know where it falls short. 

### Less Fictional Datasets

We are going to be using the same datasets Max Harlow used for this exercise. As he explains in his presentation [here](https://docs.google.com/presentation/d/1djKgqFbkYDM8fdczFhnEJLwapzmt4RLuEjXkJZpKves/edit#slide=id.g3512a0ce6b_1_22), there are a bunch of files: 
- a list of world billionaires published by Bloomberg
- a similar list published by Forbes
- a list also published by Forbes that only includes Chinese individuals
- a list published by the CIA of chiefs of state and cabinet members of foreign governments
- a list of all the people that attended the World Economic Forum conference in Davos this year
- a list of all the people and companies that have been sanctioned by the United Nations


In [11]:
## Read in the two billionaire lists (Forbes + Bloomberg)

forbes_df = pd.read_csv("sources/forbes-billionaires.csv")
bloom_df = pd.read_csv("sources/bloomberg-billionaires.csv")

Can you find out how many billionaires appear in both lists (exact matching)? 

In [12]:
forbes_df.sample(30)

name         lastName                  uri  \
635              Liu Yonghao              Liu          liu-yonghao   
1929           Djoko Susanto          Susanto        djoko-susanto   
1042              Marc Rowan            Rowan           marc-rowan   
1388           Abhay Firodia          Firodia        abhay-firodia   
348          Pallonji Mistry           Mistry      pallonji-mistry   
1991            Archie Hwang            Hwang         archie-hwang   
873               Wang Yusuo             Wang           wang-yusuo   
1692            Jiang Xuefei            Jiang         jiang-xuefei   
405              Graeme Hart             Hart          graeme-hart   
2002              Lev Leviev           Leviev           lev-leviev   
586                Jim Davis            Davis            jim-davis   
56    Don Vultaggio & family        Vultaggio        don-vultaggio   
1817              Que Wenbin              Que           que-wenbin   
812               Frank Wang             Wang           frank-wang   
2131              Lee Ho-Jin              Lee           lee-ho-jin   
615             David Siegel           Siegel       david-siegel-1   
1774             Victor Fung             Fung          victor-fung   
456             Sunil Mittal           Mittal         sunil-mittal   
1373          Henri Beaufour         Beaufour       henri-beaufour   
1101              Bill Gross            Gross           bill-gross   
2140              Liu Shaoxi              Liu           liu-shaoxi   
468       Hank & Doug Meijer           Meijer     hank-doug-meijer   
1491         Gerald Schwartz         Schwartz      gerald-schwartz   
1286          Peter Peterson         Peterson       peter-peterson   
1057      Otto Philipp Braun            Braun   otto-philipp-braun   
1478         Patrice Motsepe          Motsepe      patrice-motsepe   
543      Eva Gonda de Rivera  Gonda de Rivera  eva-gonda-de-rivera   
191      Prathip Chiravanond      Chiravanond  prathip-chiravanond   
1440            Cai Wensheng              Cai         cai-wensheng   
960       Romesh T. Wadhwani         Wadhwani    romesh-t-wadhwani   

                imageUri  worthChange  \
635          liu-yonghao       12.684   
1929       djoko-susanto      -13.901   
1042          marc-rowan       33.560   
1388       abhay-firodia       19.893   
348      pallonji-mistry      334.230   
1991              no-pic        0.000   
873           wang-yusuo        1.710   
1692              no-pic        1.539   
405          graeme-hart        0.000   
2002          lev-leviev        2.043   
586            jim-davis        0.000   
56                no-pic        0.000   
1817              no-pic          NaN   
812           frank-wang        0.000   
2131          lee-ho-jin        0.200   
615       david-siegel-1        0.000   
1774         victor-fung          NaN   
456         sunil-mittal       89.479   
1373              no-pic       34.114   
1101          bill-gross        0.000   
2140          liu-shaoxi          NaN   
468     hank-doug-meijer        0.000   
1491     gerald-schwartz        2.939   
1286      peter-peterson          NaN   
1057  otto-philipp-braun        0.000   
1478     patrice-motsepe       22.933   
543        no-pic-female        2.774   
191               no-pic       22.536   
1440              no-pic          NaN   
960    romesh-t-wadhwani        0.000   

                                           source                    industry  \
635                                  agribusiness                     Service   
1929                                 supermarkets            Fashion & Retail   
1042                               private equity     Finance and Investments   
1388                                  automobiles                  Automotive   
348                                  construction  Construction & Engineering   
1991                               semiconductors                  Technology   
873           

In [13]:
bloom_df.sample(30)

Rank                           Name Total_net_worth             Country  \
366   367       Gwendolyn Sontheim Meyer          $4.58B       United States   
264   265                    Natie Kirsh          $5.91B        South Africa   
260   261                  Tsai Eng-Meng          $5.98B              Taiwan   
223   224               Alexey Kuzmichev          $6.50B  Russian Federation   
393   394           Anders Holch Povlsen          $4.38B             Denmark   
137   138                        Lei Jun          $9.77B               China   
166   167                    Ma Jianrong          $8.22B               China   
160   161                Pam Mars-Wright          $8.37B       United States   
411   412                 Micky Jagtiani          $4.25B               India   
370   371            Herbert Johnson III          $4.54B       United States   
435   436                 B Wayne Hughes          $4.12B       United States   
454   455               Kjell Inge Rokke          $4.00B              Norway   
220   221                     Zeng Yuqun          $6.54B           Hong Kong   
373   374                    Udo Tschira          $4.54B             Germany   
258   259                 Giorgio Armani          $6.01B               Italy   
361   362                 Taylor Thomson          $4.61B              Canada   
463   464                     Zhang Yong          $3.92B           Singapore   
75     76  Charlene de Carvalho-Heineken          $14.2B         Netherlands   
205   206            Reinhold Schmieding          $6.98B       United States   
304   305              Ermirio de Moraes          $5.12B              Brazil   
236   237                      Jay Y Lee          $6.36B  Korea, Republic of   
140   141               Karl Albrecht Jr          $9.48B             Germany   
119   120                  Jim Goodnight          $10.8B       United States   
66     67                 Leonard Lauder          $14.8B       United States   
462   463                   Henry Laufer          $3.95B       United States   
390   391                     Richard Li          $4.42B           Hong Kong   
344   345                  Pierre Bellon          $4.71B              France   
16     17                  Mukesh Ambani          $42.9B               India   
48     49                    Yang Huiyan          $17.7B               China   
424   425                  Anthoni Salim          $4.17B           Indonesia   

            Industry  
366      Commodities  
264  Food & Beverage  
260  Food & Beverage  
223      Diversified  
393           Retail  
137       Technology  
166         Consumer  
160  Food & Beverage  
411           Retail  
370         Consumer  
435         Services  
454       Industrial  
220       Industrial  
373       Technology  
258         Consumer  
361  Media & Telecom  
463  Food & Beverage  
75   Food & Beverage  
205      Health Care  
304      Commodities  
236      Diversified  
140           Retail  
119       Technology  
66          Consumer  
462          Finance  
390  Media & Telecom  
344           Retail  
16            Energy  
48       Real Estate  
424      Diversified

In [14]:
billionaires_df = fpd.fuzzy_merge(forbes_df, bloom_df, left_on='name', right_on='Name')
billionaires_df.head()

name   lastName               uri imageUri  worthChange  \
0    Alexander Otto       Otto    alexander-otto   no-pic        2.120   
1     Ben Ashkenazy  Ashkenazy     ben-ashkenazy   no-pic        0.000   
2  Giovanni Ferrero    Ferrero  giovanni-ferrero   no-pic        0.000   
3       Henry Cheng      Cheng     henry-cheng-1   no-pic        3.542   
4      Henry Laufer     Laufer      henry-laufer   no-pic        0.000   

                source                 industry gender        country  \
0          real estate              Real Estate      M        Germany   
1          real estate              Real Estate      M  United States   
2  Nutella, chocolates        Food and Beverage      M          Italy   
3             property              Diversified      M      Hong Kong   
4          hedge funds  Finance and Investments      M  United States   

       timestamp  realTimeWorth  realTimeRank  realTimePosition  \
0  1547575201867      10821.927         126.0             126.0   
1  1547574901333       4000.000         499.0             499.0   
2  1547575201866      22673.165          38.0              38.0   
3  1547574901334       1334.282        1630.0            1630.0   
4  1547574901333       2000.000        1141.0            1142.0   

                                         squareImage  Rank              Name  \
0  //specials-images.forbesimg.com/imageserve/5a7...   323    Alexander Otto   
1  //specials-images.forbesimg.com/imageserve/59e...   447     Ben Ashkenazy   
2  //specials-images.forbesimg.com/imageserve/5b1...    33  Giovanni Ferrero   
3  //specials-images.forbesimg.com/imageserve/5a7...    79       Henry Cheng   
4                                                NaN   463      Henry Laufer   

  Total_net_worth        Country         Industry  
0          $4.94B        Germany      Real Estate  
1          $4.05B  United States      Real Estate  
2          $22.6B          Italy  Food & Beverage  
3          $14.1B      Hong Kong           Retail  
4          $3.95B  United States          Finance

Now, can you find the ones where the ranks aren't the same across the two datasets? What about the ones that are the same?

In [15]:
billionaires_df[billionaires_df.realTimeRank == billionaires_df.Rank].shape

(12, 19)

### Fuzzy matching with non-fictional data

In the above couple of cells, we've conducted "exact matching", i.e. the equivalent of you running a `Cmd+F`/`Ctrl+F` on your text editor. But, this is _almost_ worse as it's case sensitive, i.e. "Tom" and "tom" are treated differently. 

We've gone through some of this already, but what are the things we can ignore when it comes to name-matching? Harlow, in his presentation, identified:
- case
- title (Mr., Mrs., etc.)
- non-latin characters (é, å, ß, etc.)
- the order of the names
- non-alphanumerics (e.g. hyphenated names)

Now, you don't _have to_ ignore _anything_, but sometimes, it might make your life far easier. Other times, you'll end up with false positives and whatnot. 

The library `csvmatch`—and by extension `fuzzy_pandas`—support a bunch of the above parameters, which you can just pass in to the function. Passing in a bunch of these parameters would allow you to go from `Orbán, Viktor` to `Viktor Orban`, which is quite useful. (Again, the example's from Harlow's slides)

For this bit, we'll move on to two of the other datasets: `cia-world-leaders.csv` and `davos-attendees-2019.csv`. As always, read in the data and figure out which columns the exact match should run on. 

In [16]:
cia_world_leaders = pd.read_csv('sources/cia-world-leaders.csv')
davos_attendees = pd.read_csv('sources/davos-attendees-2019.csv')
print(f"Our CIA World Leaders df has these columns: {cia_world_leaders.columns} \
      \n The Davos attendees have these: {davos_attendees.columns}")

Our CIA World Leaders df has these columns: Index(['country', 'role', 'name'], dtype='object')       
 The Davos attendees have these: Index(['full_name', 'position_short_name', 'org_name', 'org_country'], dtype='object')


In [17]:
cia_world_leaders.sort_values('name').head(20)

country                                               role  \
384             Bangladesh                            Min. of Foreign Affairs   
499                 Belize                             Governor, Central Bank   
4721               Somalia              Min. of Public Works & Reconstruction   
4451          Saudi Arabia                                   Min. of Interior   
2625                Jordan                                               King   
4193                 Qatar                                        Deputy Amir   
4213                 Qatar                                         Prime Min.   
4205                 Qatar                                   Min. of Interior   
4194                 Qatar                       Governor, Qatar Central Bank   
5482  United Arab Emirates  Min. of Foreign Affairs and International Coop...   
5487  United Arab Emirates                 Min. of Infrastructure Development   
1809              Ethiopia                                         Prime Min.   
1732               Eritrea                               Min. of Public Works   
4891                 Sudan                      Min. of Guidance & Endowments   
1799              Ethiopia                                    Min. of Revenue   
5139              Thailand                                      Min. of Labor   
394             Bangladesh                                   Min. of Planning   
366                Bahrain      Min. of the Royal Court for Follow-Up Affairs   
4199                 Qatar                            Min. of Economy & Trade   
353                Bahrain                                    Min. of Finance   

                                                   name  
384                                   A. H. Mahmood ALI  
499                                        A. Joy GRANT  
4721                     ABAS Abdullahi Sheikh "Siraji"  
4451  ABD AL-AZIZ bin Saud bin Nayif bin Abd al-Aziz...  
2625                                        ABDALLAH II  
4193                        ABDALLAH bin Hamad Al Thani  
4213            ABDALLAH bin Nasir bin Khalifa Al Thani  
4205            ABDALLAH bin Nasir bin Khalifa Al Thani  
4194                         ABDALLAH bin Saud Al Thani  
5482                     ABDALLAH bin Zayid Al Nuhayyan  
5487             ABDULLAH bin Muhammad Belhaf al-Nuaymi  
1809                                         ABIY Ahmed  
1732                                      ABRAHA Asfaha  
4891                            ABU-BAKR Uthman Ibrahim  
1799                                      ADANECH Abebe  
5139                      ADUN Saengsingkaew, Pol. Gen.  
394                                   AHM Mostafa KAMAL  
366                    AHMAD BIN ATIYATALLAH Al Khalifa  
4199              AHMAD bin Jasim bin Muhammad Al Thani  
353   AHMAD bin Muhammad bin Hamad bin Abdallah Al K...

In [18]:
davos_attendees.sort_values('full_name').head(20)

full_name  \
1864                 Aaron Karczmer   
2532               Aaron Motsoaledi   
422              Aarthi Subramanian   
2074            Abdelkader Messahel   
897            Abdulaziz Al Judaimi   
888            Abdulaziz Al Subeaei   
35            Abdulaziz Al-Helaissi   
903             Abdulaziz Al-Jarbou   
2641               Abdulla Al Basti   
809              Abdulla Al Khalifa   
1228             Abdulla Bin Desmal   
804   Abdulla Bin Mohammed Al Thani   
2638               Abdulla Bin Touq   
2660          Abdulla Nasser Lootah   
2071              Abdullah Abdullah   
895          Abdulmajeed Al Mobarak   
663                Abdulsamad Rabiu   
2428                     Abeer Odeh   
2297               Abha Joshi-Ghani   
695                       Abid Butt   

                                    position_short_name  \
1864  Executive Vice-President; Chief Risk, Complian...   
2532                 Minister of Health of South Africa   
422              Executive Director, Board of Directors   
2074             Minister of Foreign Affairs of Algeria   
897                   Senior Vice-President, Downstream   
888                                            Chairman   
35                  Group Chief Executive Officer (GIB)   
903                               Chairman of the Board   
2641  General Secretary, Executive Council of Dubai,...   
809                             Chief Executive Officer   
1228                    Director, Commercial Operations   
804                                            Chairman   
2638  Secretary-General, Cabinet of the United Arab ...   
2660  Director-General, Federal Competitiveness and ...   
2071  Chief Executive of the Islamic Republic of Afg...   
895                             Chief Executive Officer   
663                Chairman and Chief Executive Officer   
2428  Minister of National Economy of the Palestinia...   
2297  Senior Adviser, Infrastructure, Public- Privat...   
695                             Chief Executive Officer   

                                               org_name  \
1864                                             PayPal   
2532                 Ministry of Health of South Africa   
422                           Tata Consultancy Services   
2074             Ministry of Foreign Affairs of Algeria   
897                                        Saudi Aramco   
888                     Jabal Omar Development Company    
35                          Gulf International Bank BSC   
903                 Saudi Basic Industries Corporation    
2641                         Executive Council of Dubai   
809                        Qatar National Bank Q.P.S.C.   
1228                                   DP World Limited   
804                                             Ooredoo   
2638                Cabinet of the United Arab Emirates   
2660  The Federal Competitiveness and Statistics Aut...   
2071                             Afghanistan Government   
895                                          Riyad Bank   
663                                           BUA Group   
2428  Ministry of National Economy of the Palestinia...   
2297                                     The World Bank   
695                         e2e Supply Chain Management   

                  org_country  
1864                      USA  
2532             South Africa  
422                     India  
2074                  Algeria  
897              Saudi Arabia  
888              Saudi Arabia  
35                    Bahrain  
903              Saudi Arabia  
2641     United Arab Emirates  
809                     Qatar  
1228     United Arab Emirates  
804                     Qatar  
2638     United Arab Emirates  
2660     United Arab Emirates  
2071              Afghanistan  
895              Saudi Arabia  
663                   Nigeria  
2428  Palestinian Territories  
2297                      INT  
695                  Pakistan

In [19]:
# Let's start by seeing what an exact match would look like. 
davos_df = fpd.fuzzy_merge(cia_world_leaders, davos_attendees, left_on='name', right_on='full_name')
davos_df.shape

(0, 7)

Apparently, there are no matches. What do you reckon? What _should_ the overlap between CIA's list of world leaders and Davos attendees be? 

Let's try some _fuzzy matching_, first by simply ignoring case. 

In [20]:
davos_df = fpd.fuzzy_merge(cia_world_leaders, davos_attendees, left_on='name', right_on='full_name', ignore_case=True)

In [21]:
print(davos_df.shape)
davos_df

(119, 7)


country                                               role  \
0           Algeria  Min. of Foreign Affairs & International Cooper...   
1         Argentina                          Min. of Production & Work   
2         Argentina                                Pres., Central Bank   
3           Armenia                                         Prime Min.   
4         Australia                          Min. for Defense Industry   
..              ...                                                ...   
114  United Kingdom  Sec. of State for Intl. Trade & President of t...   
115         Vietnam                           Min. of Industry & Trade   
116         Vietnam            Min. of Natural Resources & Environment   
117         Vietnam                                         Prime Min.   
118        Zimbabwe             Min. of Finance & Economic Development   

                    name            full_name  \
0    Abdelkader MESSAHEL  Abdelkader Messahel   
1             Dante SICA           Dante Sica   
2        Guido SANDLERIS      Guido Sandleris   
3        Nikol PASHINYAN      Nikol Pashinyan   
4           Steven CIOBO         Steven Ciobo   
..                   ...                  ...   
114             Liam FOX             Liam Fox   
115        Tran Tuan ANH        Tran Tuan Anh   
116         Tran Hong HA         Tran Hong Ha   
117     Nguyen Xuan PHUC     Nguyen Xuan Phuc   
118         Mthuli NCUBE         Mthuli Ncube   

                                   position_short_name  \
0               Minister of Foreign Affairs of Algeria   
1         Minister of Industry and Labour of Argentina   
2            Governor of the Central Bank of Argentina   
3            Prime Minister of the Republic of Armenia   
4            Minister of Defence Industry of Australia   
..                                                 ...   
114  Secretary of State for International Trade of ...   
115         Minister of Industry and Trade of Viet Nam   
116  Minister of Natural Resources and Environment ...   
117                         Prime Minister of Viet Nam   
118  Minister of Finance and Economic Development o...   

                                              org_name     org_country  
0               Ministry of Foreign Affairs of Algeria         Algeria  
1         Ministry of Industry and Labour of Argentina       Argentina  
2                            Central Bank of Argentina       Argentina  
3    Office of the Prime Minister of the Republic o...         Armenia  
4                   Department of Defence of Australia       Australia  
..                                                 ...             ...  
114                 Department for International Trade  United Kingdom  
115         Ministry of Industry and Trade of Viet Nam        Viet Nam  
116  Ministry of Natural Resources and Environment ...        Viet Nam  
117               Office of the Government of Viet Nam        Viet Nam  
118  Ministry of Finance and Economic Development o...        Zimbabwe  

[119 rows x 7 columns]

OK, we have more matches, but this is also pretty boring. There's nothing super-smart about ignoring cases to get matches. Your word processors have been doing that for _decades_. 

But, now, let's start adding some of our other parameters discussed above, and see what happens

In [22]:
davos_df = fpd.fuzzy_merge(cia_world_leaders, davos_attendees, left_on='name', right_on='full_name', 
                           ignore_case=True, 
                           ignore_order_words=True,
                           ignore_nonalpha=True,
                           ignore_titles=True
                          )

In [23]:
print(davos_df.shape)
davos_df

(121, 7)


country                                               role  \
0           Algeria  Min. of Foreign Affairs & International Cooper...   
1         Argentina                          Min. of Production & Work   
2         Argentina                                Pres., Central Bank   
3           Armenia                                         Prime Min.   
4         Australia                          Min. for Defense Industry   
..              ...                                                ...   
116  United Kingdom  Sec. of State for Intl. Trade & President of t...   
117         Vietnam                           Min. of Industry & Trade   
118         Vietnam            Min. of Natural Resources & Environment   
119         Vietnam                                         Prime Min.   
120        Zimbabwe             Min. of Finance & Economic Development   

                    name            full_name  \
0    Abdelkader MESSAHEL  Abdelkader Messahel   
1             Dante SICA           Dante Sica   
2        Guido SANDLERIS      Guido Sandleris   
3        Nikol PASHINYAN      Nikol Pashinyan   
4           Steven CIOBO         Steven Ciobo   
..                   ...                  ...   
116             Liam FOX             Liam Fox   
117        Tran Tuan ANH        Tran Tuan Anh   
118         Tran Hong HA         Tran Hong Ha   
119     Nguyen Xuan PHUC     Nguyen Xuan Phuc   
120         Mthuli NCUBE         Mthuli Ncube   

                                   position_short_name  \
0               Minister of Foreign Affairs of Algeria   
1         Minister of Industry and Labour of Argentina   
2            Governor of the Central Bank of Argentina   
3            Prime Minister of the Republic of Armenia   
4            Minister of Defence Industry of Australia   
..                                                 ...   
116  Secretary of State for International Trade of ...   
117         Minister of Industry and Trade of Viet Nam   
118  Minister of Natural Resources and Environment ...   
119                         Prime Minister of Viet Nam   
120  Minister of Finance and Economic Development o...   

                                              org_name     org_country  
0               Ministry of Foreign Affairs of Algeria         Algeria  
1         Ministry of Industry and Labour of Argentina       Argentina  
2                            Central Bank of Argentina       Argentina  
3    Office of the Prime Minister of the Republic o...         Armenia  
4                   Department of Defence of Australia       Australia  
..                                                 ...             ...  
116                 Department for International Trade  United Kingdom  
117         Ministry of Industry and Trade of Viet Nam        Viet Nam  
118  Ministry of Natural Resources and Environment ...        Viet Nam  
119               Office of the Government of Viet Nam        Viet Nam  
120  Ministry of Finance and Economic Development o...        Zimbabwe  

[121 rows x 7 columns]

Right, 19 more results. Baby steps, but at least steps in the right direction. Now, let's start using the more _intelligent_ algorithms in place—this one named after a Russian mathematician: Levenshtein. 

All *Levenshtein* does is look at how many characters are different between two inputs? For example:

In [24]:
from jellyfish._jellyfish import damerau_levenshtein_distance

damerau_levenshtein_distance("Évry", "Every")

2

Let's quickly use the algorithm directly to see the output. The above cell imports something called `jellyfish`, which is another package that `csvmatch` uses. Typically, you wouldn't call the function directly (you could if you wanted to), but this is just to give you an idea of how the algorithm works. 

Right, now let's use this with our above data, and see if we have better luck. 

Remember: the threshold specified by us is 0.6, and it's calculated by: `1-(distance/max(value1, value2))`. In the case of `Évry` and `Every` above, our calculation would be:

`1 - (2/5)` = `3/5` = `0.6`

So, in this case, the two would lead to a fuzzy match.

In [31]:
davos_df = fpd.fuzzy_merge(cia_world_leaders, davos_attendees, left_on='name', right_on='full_name', 
                           ignore_case=True, 
                           ignore_order_words=True,
                           ignore_nonalpha=True,
                           ignore_titles=True,
                           method='levenshtein'
                          )

In [32]:
davos_df.shape

(1419, 7)

**WAIT, WHAT?!** Have we just gone from 138 matches to 1952? Is that overtly optimistic? 

In [27]:
davos_df.sample(50)

country                                               role  \
1358        United Kingdom  Sec. of State for Environment, Food, & Rural A...   
1253                 Sudan                                      Attorney Gen.   
68               Australia                                    Dep. Prime Min.   
1391               Vietnam                                              Pres.   
692           Korea, North                              Member, SPA Presidium   
638                  Japan          Min. in Charge of Regional Revitalization   
305                  China                                Min. of Agriculture   
311                  China          Min. of Human Resources & Social Security   
874       Marshall Islands                                    Min. of Justice   
1024               Nigeria                                Min. of Environment   
731                 Kosovo                           Min. of Trade & Industry   
1020               Nigeria                                    Min. of Defense   
1202              Slovakia            Min. of Agriculture & Rural Development   
110              Australia                      Min. for Regional Development   
468                  Gabon  Min. of Interior, Security, & Public Hygiene i...   
1198              Slovakia     Dep. Prime Min. for Investment and Information   
875       Marshall Islands                                    Min. of Justice   
229                 Brazil                               Ambassador to the US   
29     Antigua and Barbuda           Min. of Education, Science, & Technology   
44               Argentina                Min. of Health & Social Development   
1061      Papua New Guinea                                   Min. for Justice   
676                  Kenya                    Governor, Central Bank of Kenya   
472                  Gabon                                      Min. of State   
753                Lebanon                         Min. of Telecommunications   
895                Moldova                  Dep. Prime Min. for Reintegration   
97               Australia                            Min. for Human Services   
631                  Japan          Min. for Economic Cooperation With Russia   
924                  Nauru                            Min. Asst. to the Pres.   
816                 Malawi        Min. of Foreign Affairs & Intl. Cooperation   
645                  Japan                  Min. of State for Gender Equality   
369                Croatia       Permanent Representative to the UN, New York   
912             Mozambique                               Ambassador to the US   
902               Mongolia                    Min. of Mining & Heavy Industry   
1103                  Peru                                              Pres.   
802             Luxembourg                                      Min. of State   
1005           New Zealand                                    Min. of Revenue   
1330  United Arab Emirates                 Min. of Infrastructure Development   
1310               Tunisia                                   Min. of Commerce   
325                  China                             Premier, State Council   
96               Australia                            Min. for Human Services   
77               Australia                          Min. for Defense Industry   
217                Bermuda                              Min. of Public Safety   
1185             Singapore             Chmn., Monetary Authority of Singapore   
355             Costa Rica                             Min. of Women's Issues   
710           Korea, North       Permanent Representative to the UN, New York   
1121                 Qatar                            Min. of Foreign Affairs   
1371        United Kingdom                        Sec. of State for Transport   
1163                Serbia                            Min. of Energy & Mining   
269                 Canada                  Leader

Why, yes. Yes, it is. This is why you _always_ confirm what an algorithm does. Right, maybe let's bump up our threshold and see what happens. 

In [28]:
davos_df = fpd.fuzzy_merge(cia_world_leaders, davos_attendees, left_on='name', right_on='full_name', 
                           ignore_case=True, 
                           ignore_order_words=True,
                           ignore_nonalpha=True,
                           ignore_titles=True,
                           method='levenshtein',
                           threshold=0.8
                          )

In [29]:
davos_df.shape

(182, 7)

OK, that seems more reasonable. Let's sanity check.

In [30]:
davos_df.sample(50)

country                                               role  \
27           Canada                            Min. of Foreign Affairs   
104      Luxembourg                                         Prime Min.   
32       Costa Rica                                              Pres.   
159         Tunisia  Min. of Development, Investment & Internationa...   
28            China                                    Chief Executive   
97    Liechtenstein            Min. of General Govt. Affairs & Finance   
163          Uganda                                              Pres.   
147    South Africa                       Min. of Economic Development   
155     Switzerland                                 Federal Chancellor   
107        Malaysia                     Min. of Intl. Trade & Industry   
172  United Kingdom                Sec. of State for Intl. Development   
80            Italy                            Min. of Foreign Affairs   
8         Australia                            Min. for Public Service   
151     Switzerland  Chief, Federal Dept. of Defense, Civil Protect...   
88       Kazakhstan                                    Dep. Prime Min.   
10          Austria                                         Chancellor   
72          Ireland               Min. for Public Expenditure & Reform   
9         Australia              Min. for Trade, Investment, & Tourism   
11          Austria    Min. for Europe, Integration, & Foreign Affairs   
41            Egypt    Min. of Housing, Utilities, & Urban Development   
74          Ireland                             Taoiseach (Prime Min.)   
137    Saudi Arabia                                    Min. of Finance   
143       Singapore               Min. for Communication & Information   
122         Nigeria              Min. of Industry, Trade, & Investment   
100      Luxembourg                     Min. of Communications & Media   
24         Cambodia              Min. of Public Works & Transportation   
53           France                          Min. of Economy & Finance   
101      Luxembourg                                    Min. of Culture   
98    Liechtenstein                                         Prime Min.   
113     Netherlands                              Min. for Medical Care   
130          Poland                            Min. of Foreign Affairs   
59          Germany                                    Min. of Defense   
120     New Zealand                         Min. of Sport & Recreation   
153     Switzerland            Chief, Federal Dept. of Foreign Affairs   
180        Zimbabwe              Min. of Foreign Affairs & Intl. Trade   
171  United Kingdom             Sec. of State for Health & Social Care   
105       Macedonia                                         Prime Min.   
178         Vietnam                                         Prime Min.   
112     Netherlands   Min. for Foreign Trade & Development Cooperation   
31            China                                   Sec. for Finance   
13          Bahrain         Min. of Transportation & Telecommunication   
119     New Zealand                                    Min. of Finance   
87           Jordan  Min. of State for Investment Affairs & Chmn. o...   
52           France                  Min. for Europe & Foreign Affairs   
21         Bulgaria                                         Prime Min.   
38          Denmark                                   Min. for Finance   
140       Singapore             Chmn., Monetary Authority of Singapore   
5           Armenia                                         Prime Min.   
48             Fiji       Permanent Representative to the UN, New York   
1           Algeria  Min. of Foreign Affairs & International Cooper...   

                               name                 full_name  \
27                Chrystia FREELAND         Chrystia Freeland   
104                   Xavier BETTEL             Xavier Bettel   
32          Carlos ALVARADO Quesada   Carl

Next question, one for you guys to do:

Which names from the CIA world leaders list are also on the Forbes billionaires list?

Who can get the best result?

In [33]:
ciabill_df = fpd.fuzzy_merge(cia_world_leaders, forbes_df, left_on='name', right_on='name', 
                           ignore_case=True, 
                           ignore_order_words=True,
                           ignore_nonalpha=True,
                           ignore_titles=True,
                           method='levenshtein',
                           threshold=0.8
                          )

In [34]:
ciabill_df.shape

(12, 17)

So in what scenarios do you reckon Levenshtein will perform badly? 

Next up: **metaphone**. 

Metaphone's great for names which sound similar, which wouldn't be caught by Levenshtein. It's especially handy when you're working with transcript data. But, it too comes with its pitfalls. 

Let's look at an example and then discuss what the possible pitfalls could be. 

Which names from the CIA world leaders list are also on the United Nations sanctions list?

In [35]:
sanctions_df = pd.read_csv("sources/un-sanctions.csv")

In [37]:
ciasanct_df = fpd.fuzzy_merge(cia_world_leaders, sanctions_df, left_on='name', right_on='name', 
                           method='metaphone'
                          )

In [38]:
ciasanct_df.shape

(18, 22)

In [ ]:
## How does this compare to our other algorithms (ignore case)? 


In [ ]:
# ciasanct2_df.shape

In [ ]:
## How does this compare to our other algorithms (levenshtein)? 
ciasanct3_df = fpd.fuzzy_merge(cia_world_leaders, sanctions_df, left_on='name', right_on='name', 
                           ignore_case=True, 
                           ignore_order_words=True,
                           ignore_nonalpha=True,
                           ignore_titles=True,
                           method='levenshtein',
                           threshold=0.8
                          )

In [ ]:
ciasanct3_df.shape

Finally, we get to **Bilenko**. And, yes, this uses machine learning, where you train your own data. So, now you have human *smarts* being involved in the process of matching up names across documents. 

Let's look at an example: 

Which names from the CIA world leaders list are also on the Davos attendees list?

In [ ]:
ciasanct4_df = fpd.fuzzy_merge(cia_world_leaders, sanctions_df, left_on='name', right_on='name', 
                           method='bilenko'
                          )


Answer questions as follows:
 y - yes
 n - no
 s - skip
 f - finished

name: OMAR bin Sultan al-Ulama

name: OMAR HAMMAMI

Do these records refer to the same thing? [y/n/s/f] 

n



name: Imran KHAN

name: PEJMAN INDUSTRIAL SERVICES CORPORATION

Do these records refer to the same thing? [y/n/s/f] 

n



name: Milena HARITO

name: OCEAN MARITIME MANAGEMENT COMPANY, LIMITED (OMM)

Do these records refer to the same thing? [y/n/s/f] 

n



name: AMIR Aman

name: AMIN MUHAMMAD UL HAQ

Do these records refer to the same thing? [y/n/s/f] 

n



name: Paul BIYOGHE MBA

name: CENTRAL MILITARY COMMISSION OF THE WORKERS’ PARTY OF KOREA (CMC)

Do these records refer to the same thing? [y/n/s/f] 

n



name: Muhammad bin Hamad bin Sayf al-RUMHI

name: Muhammad Bahrum Naim

Do these records refer to the same thing? [y/n/s/f] 

n



name: Cecilia PEREZ

name: RICARDO PEREZ AYERAS

Do these records refer to the same thing? [y/n/s/f] 

n



name: Yeafesh OSMAN

name: STATE ENTERPRISE FOR MARKETING EQUIPMENT AND MAINTENANCE

Do these records refer to the same thing? [y/n/s/f] 

n



name: MUHAMMAD BIN RASHID Al Maktum

name: MOHAMMAD ESLAMI

Do these records refer to the same thing? [y/n/s/f] 

n



name: Peter MUNYA

name: STATE BATTERY MANUFACTURING ESTABLISHMENT

Do these records refer to the same thing? [y/n/s/f] 

n



name: CHOE Pu Il

name: CHOE CHUN YONG

Do these records refer to the same thing? [y/n/s/f] 

n



name: WEI Fenghe

name: WEIHAI WORLD-SHIPPING FREIGHT

Do these records refer to the same thing? [y/n/s/f] 

n



name: Ravi YERRIGADOO

name: RAFIDAIN STATE ORGANIZATION FOR IRRIGATION PROJECTS

Do these records refer to the same thing? [y/n/s/f] 

n



name: Anda CAKSA

name: STATE ORGANIZATION FOR AGRICULTURAL MECHANIZATION AND AGRICULTURAL SUPPLIES

Do these records refer to the same thing? [y/n/s/f] 

n



name: Ebrima MBALLOW

name: STATE ESTABLISHMENT FOR AGRICULTURAL MARKETING

Do these records refer to the same thing? [y/n/s/f] 

n



name: RI Su Yong

name: RI SU YONG

Do these records refer to the same thing? [y/n/s/f] 

y



name: Ivo VALENTE

name: NATIONAL ENTERPRISE FOR EQUIPMENT MARKETING AND MAINTENANCE

Do these records refer to the same thing? [y/n/s/f] 

n



name: MOHAMMED VI

name: MOHAMMED AL GHABRA

Do these records refer to the same thing? [y/n/s/f] 

n



name: Erna SOLBERG

name: STATE ENTERPRISE FOR LEATHER INDUSTRIES

Do these records refer to the same thing? [y/n/s/f] 

n



name: Muhammad al-AMARI

name: Muhammad Bahrum Naim

Do these records refer to the same thing? [y/n/s/f] 

n



name: CHOE Sung Ho

name: CHOE HWI

Do these records refer to the same thing? [y/n/s/f] 

n



name: ZHONG Shan

name: SHEN ZHONG INTERNATIONAL SHIPPING

Do these records refer to the same thing? [y/n/s/f] 

y



name: CHOE Yong Rim

name: CHOE HWI

Do these records refer to the same thing? [y/n/s/f] 

n



name: Talal ARSLAN

name: NATIONAL CENTRE FOR ENGINEERING AND ARCHITECTURAL CONSULTANCY

Do these records refer to the same thing? [y/n/s/f] 

n



name: ABDULLAH bin Muhammad Belhaf al-Nuaymi

name: ABDULLAH ANSHORI

Do these records refer to the same thing? [y/n/s/f] 

n



name: KANG Kyung-wha

name: KANG RYONG

Do these records refer to the same thing? [y/n/s/f] 

n



name: SHEN Jong-chin

name: SHEN ZHONG INTERNATIONAL SHIPPING

Do these records refer to the same thing? [y/n/s/f] 

n



name: Steadroy BENJAMIN

name: VOCATIONAL TRAINING CENTRE FOR ENGINEERING AND METALLIC INDUSTRIES

Do these records refer to the same thing? [y/n/s/f] 

n



name: MIAO Wei

name: M23

Do these records refer to the same thing? [y/n/s/f] 

n



name: KANG Myong Chol

name: KANG RYONG

Do these records refer to the same thing? [y/n/s/f] 

n



name: Henry PUNA

name: TARKHAN TAYUMURAZOVICH BATIRASHVILI

Do these records refer to the same thing? [y/n/s/f] 

n



name: JANG Chol

name: JANG BOM SU

Do these records refer to the same thing? [y/n/s/f] 

n



name: Bruno LE MAIRE

name: TOUS POUR LA PAIX ET LE DEVELOPPEMENT (NGO)

Do these records refer to the same thing? [y/n/s/f] 

n



name: Borut PAHOR

name: STATE CONTRACTING BUILDINGS COMPANY

Do these records refer to the same thing? [y/n/s/f] 

n



name: ALI Abdu

name: ALI MUSA AL-SHAWAKH

Do these records refer to the same thing? [y/n/s/f] 

n



name: Muhammad al-AHMED

name: Muhammad Bahrum Naim

Do these records refer to the same thing? [y/n/s/f] 

n



name: Arben ADEMI

name: SECOND ACADEMY OF NATURAL SCIENCES

Do these records refer to the same thing? [y/n/s/f] 

n



name: HAN ZAW

name: HAN YU-RO

Do these records refer to the same thing? [y/n/s/f] 

n



name: Muhammad bin Mazyad al-TUWAYJRI

name: Muhammad Bahrum Naim

Do these records refer to the same thing? [y/n/s/f] 

n



name: CHANG Po-ya

name: CHANG AN SHIPPING & TECHNOLOGY

Do these records refer to the same thing? [y/n/s/f] 

nn


Do these records refer to the same thing? [y/n/s/f] 

n



name: WANG Menghui

name: SINGWANG ECONOMICS AND TRADING GENERAL CORPORATION

Do these records refer to the same thing? [y/n/s/f] 

n



name: SALIM al-Abdallah al-Jabir al-Sabah

name: SALIM KONY

Do these records refer to the same thing? [y/n/s/f] 

n



name: Omar HILALE

name: MUSA HILAL ABDALLA

Do these records refer to the same thing? [y/n/s/f] 

n



name: Denis NAUGHTEN

name: STATE ESTABLISHMENT FOR SLAUGHTERING HOUSES

Do these records refer to the same thing? [y/n/s/f] 

n



name: Piyush GOYAL

name: AL-HARAMAIN & AL MASJED AL-AQSA CHARITY FOUNDATION

Do these records refer to the same thing? [y/n/s/f] 

n



name: Colin JORDAN

name: KEVIN JORDAN AXEL

Do these records refer to the same thing? [y/n/s/f] 

n



name: Ahmad bin Abdallah al-MAHMUD

name: Ahmad Iman Ali

Do these records refer to the same thing? [y/n/s/f] 

n



name: Ahmad bin Muhammad bin Salim al-FUTAISI

name: Ahmad Iman Ali

Do these records refer to the same thing? [y/n/s/f] 

n



name: MOHAMMED VI

name: ABOU MOHAMED AL ADNANI

Do these records refer to the same thing? [y/n/s/f] 

n



name: Pierre RAFFOUL

name: فيض

Do these records refer to the same thing? [y/n/s/f] 

n



name: Aysha MOHAMMED

name: TARAD MOHAMMAD ALJARBA

Do these records refer to the same thing? [y/n/s/f] 

n



name: Darmin NASUTION

name: STATE ENTERPRISE FOR TEXTILE AND SPINNING PRODUCTS IMPORTING AND DISTRIBUTION

Do these records refer to the same thing? [y/n/s/f] 

n



name: Ian DOUGLAS

name: IRUTA DOUGLAS MPAMO

Do these records refer to the same thing? [y/n/s/f] 

n



name: Dharmendra PRADHAN

name: CHEMICAL, PETROCHEMICAL, MECHANICAL AND METALURICAL TRAINING CENTRE

Do these records refer to the same thing? [y/n/s/f] 

n



name: CHOE Sung Ho

name: CHOE SONG IL

Do these records refer to the same thing? [y/n/s/f] 

n



name: MUHAMMAD BIN SALMAN bin Abd al-Aziz Al Saud

name: MAALIM SALMAN

Do these records refer to the same thing? [y/n/s/f] 

n



name: Muhammad al-JABRI

name: Muhammad Bahrum Naim

Do these records refer to the same thing? [y/n/s/f] 

n



name: MUJAHID bin Yusof

name: MUJAHIDIN INDONESIAN TIMUR (MIT)

Do these records refer to the same thing? [y/n/s/f] 

n



name: Muhammad HASHIM

name: Muhammad Bahrum Naim

Do these records refer to the same thing? [y/n/s/f] 

n



name: CHO Myoung-gyon

name: CHO IL U

Do these records refer to the same thing? [y/n/s/f] 

n



name: AHMAD bin Muhammad bin Hamad bin Abdallah Al Khalifa

name: AHMAD VAHID DASTJERDI

Do these records refer to the same thing? [y/n/s/f] 

n



name: CHEA CHANTO

name: CHEMICAL, PETROCHEMICAL, MECHANICAL AND METALURICAL TRAINING CENTRE

Do these records refer to the same thing? [y/n/s/f] 

In [ ]:
ciasanct4_df.shape